File: 14-Facade/src/go/home_theater_facade.go
package main

import "fmt"

// HomeTheaterFacade — simple interface over the home theater subsystem.
// It holds the subsystem components and sequences their calls.
type HomeTheaterFacade struct {
	projector Projector
	screen    Screen
	amplifier Amplifier
	lights    TheaterLights
	dvdPlayer DvdPlayer
}

func NewHomeTheaterFacade(projector Projector, screen Screen, amplifier Amplifier,
	lights TheaterLights, dvdPlayer DvdPlayer) *HomeTheaterFacade {
	return &HomeTheaterFacade{
		projector: projector,
		screen:    screen,
		amplifier: amplifier,
		lights:    lights,
		dvdPlayer: dvdPlayer,
	}
}

func (h *HomeTheaterFacade) WatchMovie(movie string) {
	fmt.Println("Get ready to watch a movie...")
	h.lights.Dim(10)
	h.screen.Down()
	h.projector.On()
	h.projector.WideScreenMode()
	h.amplifier.On()
	h.amplifier.SetVolume(8)
	h.dvdPlayer.On()
	h.dvdPlayer.Play(movie)
}

func (h *HomeTheaterFacade) EndMovie() {
	fmt.Println("Shutting movie theater down...")
	h.dvdPlayer.Stop()
	h.dvdPlayer.Off()
	h.amplifier.Off()
	h.projector.Off()
	h.screen.Up()
	h.lights.On()
}

File: 14-Facade/src/rust/home_theater_facade.rs
use crate::projector::Projector;
use crate::screen::Screen;
use crate::amplifier::Amplifier;
use crate::theater_lights::TheaterLights;
use crate::dvd_player::DvdPlayer;

/// Facade — owns the subsystem components and sequences their calls.
pub struct HomeTheaterFacade {
    projector: Projector,
    screen: Screen,
    amplifier: Amplifier,
    lights: TheaterLights,
    dvd_player: DvdPlayer,
}

impl HomeTheaterFacade {
    pub fn new(projector: Projector, screen: Screen, amplifier: Amplifier,
               lights: TheaterLights, dvd_player: DvdPlayer) -> Self {
        HomeTheaterFacade { projector, screen, amplifier, lights, dvd_player }
    }

    pub fn watch_movie(&self, movie: &str) {
        println!("Get ready to watch a movie...");
        self.lights.dim(10);
        self.screen.down();
        self.projector.on();
        self.projector.wide_screen_mode();
        self.amplifier.on();
        self.amplifier.set_volume(8);
        self.dvd_player.on();
        self.dvd_player.play(movie);
    }

    pub fn end_movie(&self) {
        println!("Shutting movie theater down...");
        self.dvd_player.stop();
        self.dvd_player.off();
        self.amplifier.off();
        self.projector.off();
        self.screen.up();
        self.lights.on();
    }
}
